source/extensions/filters/http/cdn_loop/filter.cc

```diff
@@ -0,0 +1,48 @@
+#include "extensions/filters/http/cdn_loop/filter.h"
+
+#include "envoy/http/codes.h"
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+
+#include "common/common/statusor.h"
+
+#include "extensions/filters/http/cdn_loop/utils.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CdnLoop {
+
+namespace {
+const Http::LowerCaseString CdnLoopHeaderName("CDN-Loop");
+constexpr absl::string_view ParseErrorMessage = "Invalid CDN-Loop header in request.";
+constexpr absl::string_view ParseErrorDetails = "invalid_cdn_loop_header";
+constexpr absl::string_view LoopDetectedMessage = "The server has detected a loop between CDNs.";
+constexpr absl::string_view LoopDetectedDetails = "cdn_loop_detected";
+} // namespace
+
+Http::FilterHeadersStatus CdnLoopFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                       bool /*end_stream*/) {
+  if (const Http::HeaderEntry* header_entry = headers.get(CdnLoopHeaderName);
```

Comment:
Oops - sorry I missed this on first pass, but because the header isn't inlined, I think you want HeaderUtility::getAllOfHeader, and maybe a utility to wrapper header instances and counts?

Reply:
Hmm.  If the header isn't inlined, the appendCopy on line 41 is also going to be buggy since appendCopy will modify only the first header, not the last header as required by RFC 8586.

If we're going to make it an inline header, I'm happy to keep the current implementation.  If not, I'll see what I can re-work.

Reply:
From an offline suggestion from @alyssawilk , I switched to using CustomInlineHeaders, solving this bug!